### PR TITLE
Add deflection to custom `__lt__` and add `functools.total_ordering`

### PR DESCRIPTION
**Pull request by Karrenbelt**
## Proposed changes

Adding deflection and `functools.total_ordering` to custom data types with a `__lt__` implementation. `PackageVersion`, `PublicId` and `PackageId`.

Note that string casting is still using in `__lt__` comparison; in that sense this PR is step 1 out of 2. 

## Fixes

If it fixes a bug or resolves a feature request, be sure to link to that issue.

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes and CI passes too
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by Karrenbelt** [comment]
current remaining failures:
```
FAILED tests/test_configurations/test_base.py::test_public_id_comparator_when_author_is_different
FAILED tests/test_configurations/test_base.py::test_public_id_comparator_when_name_is_different
FAILED tests/test_configurations/test_data_types.py::test_self_type_comparison[Dependency]
FAILED tests/test_configurations/test_data_types.py::test_package_version_comparison
FAILED tests/test_configurations/test_data_types.py::test_public_id_comparison
```

the first two I'm addressing here in this PR, I need to change the error message and test to check for `TypeError` rather than `ValueError`.  
the third is an issue with the test, `Dependency` does not have a `__lt__` method so should not be tested.
The last two relate the fact that we compare strings instead of tuples, where 
```
>>> ("1.2.0" < "1.10.0") == (VersionInfoClass.parse("1.2.0") < VersionInfoClass.parse("1.10.0"))
False
```
to be addressing in a subsequent PR

File: tests/test_configurations/test_data_types.py [hunk 1]
```diff
@@ -61,14 +61,12 @@ def all_comparisons_operations_equal(pair_a, pair_b) -> bool:
...
     """Test comparison to self"""
 
     self = st.from_type(self_type).example()
-    copy_self = copy.deepcopy(self)
-    other = st.from_type(self_type).example()
-    assert self == copy_self
+    other = copy.deepcopy(self)
     for f in COMPARISON_OPERATORS:
         assert isinstance(f(self, other), bool)
         assert isinstance(f(other, self), bool)
```
Comment (Karrenbelt) [inline tests/test_configurations/test_data_types.py:72]: we should copy, we don't want to test raising a `TypeError` here when certain fields (e.g. author and or name on `PublicId`) are not identical. That is done in 
https://github.com/valory-xyz/open-aea/blob/75549e587225431ecb6032c3306146f2f3b65f02/tests/test_configurations/test_data_types.py#L75-L98

**Comment by Karrenbelt** [comment]
two remaining failures are newly introduced tests in the base PR:
```
FAILED tests/test_configurations/test_data_types.py::test_package_version_comparison
FAILED tests/test_configurations/test_data_types.py::test_public_id_comparison
```

Delegating comparison to `PackageVersion._version` for all classes, however - which would be the last step required to address the issues raised in the base PR #428 - will break e.g. the DependencyTree because packages with different author and name can no longer be sorted. This will be dealt with in subsequent PRs
https://github.com/valory-xyz/open-aea/blob/45e1c7ae65b78d9477a6f89d3e6f91c7fed64b35/aea/helpers/dependency_tree.py#L248-L256

File: aea/configurations/data_types.py [hunk 2]
```diff
@@ -461,37 +456,14 @@ def __eq__(self, other: Any) -> bool:
...
-        if (
-            isinstance(other, PublicId)
-            and self.author == other.author
-            and self.name == other.name
-        ):
-            return self.package_version < other.package_version
-        raise ValueError(
-            "The public IDs {} and {} cannot be compared. Their author or name attributes are different.".format(
-                self, other
+        """Compare with another object."""
```
Comment (DavidMinarsch) [inline aea/configurations/data_types.py:459]: The old docstring was nice?

Comment (Karrenbelt) [reply aea/configurations/data_types.py:459]: bloat imo

**Review by 0xArdi — APPROVED** [review]

**Review by DavidMinarsch — APPROVED** [review]

**Pull request merged**